### PR TITLE
docs, tests and the build

### PR DESCRIPTION
**Pull request by drozdziak1**
Hello again!

I've finally managed to reach the bottom of my dependency issues - turns out there were wheel-only packages for `base58` and `morphys`. I've found the need to use separate `requirements.txt` files very confusing, so I decided to make the whole thing installable with a simple `make install` or `pip install .`. I made `tox` aware of that change, which makes the current test pipeline like so:
travis (or You) runs `tox` -> `tox` installs dependencies via `pip install .` -> `tox` runs `setup.py test` -> `setup.py` runs pytest -> pytest executes the tests and yields results

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -34,24 +34,24 @@ clean-build: ## remove build artifacts
...
-	find . -name '*.pyo' -exec rm -f {} +
-	find . -name '*~' -exec rm -f {} +
+	find . -name '*.pyc' -exec rm -fr {} +
+	find . -name '*.pyo' -exec rm -fr {} +
+	find . -name '*~' -exec rm -fr {} +
 	find . -name '__pycache__' -exec rm -fr {} +
 
 clean-test: ## remove test and coverage artifacts
-	rm -f .coverage
+	rm -fr .coverage
```
Comment (dhruvbaldawa) [inline Makefile:46]: 👍 thanks for this

File: Makefile [hunk 2]
```diff
@@ -85,4 +85,4 @@ dist: clean ## builds source and wheel package
 	ls -l dist
 
 install: clean ## install the package to the active Python's site-packages
-	python setup.py install
+	pip install .
```
Comment (dhruvbaldawa) [inline Makefile:88]: +1

File: dag/dag.py [hunk 3]
```diff
@@ -53,7 +54,8 @@ def size(self):
         return self._size
 
     @classmethod
-    def create(cls, data, links=None, hash_algorithm='sha2-256', serializer=None):
+    def create(cls, data, links=None, hash_algorithm='sha2-256',
```
Comment (dhruvbaldawa) [inline dag/dag.py]: I would set the max line length to 120 chars if that is fine. If it isn't for multiline constructs I prefer doing something like:
```python
def create(
    cls, 
    data, 
    links=None, 
    hash_algorithm='sha2-256', 
    serializer=None
):
```
so that its readable

File: requirements_dev.txt [hunk 4]
```diff
@@ -1,12 +0,0 @@
-pip==9.0.1
```
Comment (dhruvbaldawa) [inline requirements_dev.txt]: We need this file for development setup, extra things that you need to install for developing. Hence, the `_dev` suffix. This includes linters and testing utilities. So, I'd recommend to not delete this file.

File: setup.py [hunk 5]
```diff
@@ -12,9 +12,10 @@
     history = history_file.read()
 
 requirements = [
-    'morphys==1.0',
-    'pymultihash==0.8.2',
-    'pyrsistent==0.13.0',
+    'base58',
```
Comment (dhruvbaldawa) [inline setup.py]: Please pin these dependencies to their specific versions. It is important as we don't want our code to break when any of our dependencies update their package with backward incompatible change or bugs.

File: tox.ini [hunk 6]
```diff
@@ -15,11 +15,10 @@ commands=flake8 dag
 [testenv]
 setenv =
     PYTHONPATH = {toxinidir}
-deps =
-    -r{toxinidir}/requirements_dev.txt
+
+deps=.
```
Comment (dhruvbaldawa) [inline tox.ini]: Can you please revert this change, we need to install requirements_dev over here for installing pytest, flake8 and other utilities

File: tox.ini [hunk 7]
```diff
@@ -15,11 +15,10 @@ commands=flake8 dag
...
 commands =
-    pip install -U pip
-    py.test --basetemp={envtmpdir}
+    python {toxinidir}/setup.py test
```
Comment (dhruvbaldawa) [inline tox.ini]: Can you please revert this change, I prefer using pytest and this is important on CI servers to install the latest pip and then run the test command

**Review by dhruvbaldawa — CHANGES_REQUESTED** [review]
Sorry for the delayed review :( Let me know if you need any clarifications

File: dag/dag.py (hunk 3 above)
Comment (drozdziak1) [reply dag/dag.py]: Fine by me, 80 chars is just a habit of mine. Do you happen to know a tidy way to integrate `setup.cfg` flake8 entries with vim?

File: requirements_dev.txt (hunk 4 above)
Comment (drozdziak1) [reply requirements_dev.txt]: I really wish we could define such things in `test_requires` in `setup.py`, since both linting and testing are Quality Assurance aspects of writing the code, which (in my opinion) makes those *test requirements*. I can see however that using `test_requires` again makes using wheel-only test dependencies problematic.

I think a nice compromise would be to have `test_requires` in `setup.py` use a `requirements_test.txt` file so that it's both installable with `pip` and usable for `setup.py`

File: setup.py (hunk 5 above)
Comment (drozdziak1) [reply setup.py]: Okay, but at the same time maybe let's hook up a dependency updater so that we can immediately know whether new dependencies would break the project (e.g. from Travis' build results).

File: tox.ini (hunk 6 above)
Comment (drozdziak1) [reply tox.ini]: Tell me what you think about the `requirements_test.txt` approach I presented 2 comments above (we could have both `-r{toxinidir}/requirements_test.txt` and `.` in tox for testing and project dependencies, respectively)

File: tox.ini (hunk 7 above)
Comment (drozdziak1) [reply tox.ini]: But doesn't `tox` create a virtualenv for each test environment? With  `pip` installed by default?
At least `virtualenv` does that by default unless you forbid it. 

When it comes to calling `pytest` directly, I thought it would make sense to make use of the `test` target alias set in `setup.cfg` (hence `python {toxinidir}/setup.py test`) so that if we decide to switch test frameworks for whatever reason, tox would still work and there would be one file less to visit for updating the build.

**Comment by drozdziak1** [comment]
Thank you for reviewing this PR :+1: I'll take a look at your latest changes as soon as I get back home. I look forward to learning your thoughts on my comments above. Have a nice afternoon!

File: setup.py [hunk 8]
```diff
@@ -11,20 +11,18 @@
 with open('HISTORY.rst') as history_file:
     history = history_file.read()
 
+with open('requirements_test.txt') as test_requirements_file:
```
Comment (drozdziak1) [inline setup.py]: Here I'm opeining the requirements file so that we can avoid redundant dependency lists

File: tox.ini [hunk 9]
```diff
@@ -15,11 +15,13 @@ commands=flake8 dag
 [testenv]
 setenv =
     PYTHONPATH = {toxinidir}
-deps =
-    -r{toxinidir}/requirements_dev.txt
+
+deps=
+    .
```
Comment (drozdziak1) [inline tox.ini]: The `.` is to make sure that wheel-only dependencies get installed before `setup.py` starts looking for them

**Comment by drozdziak1** [comment]
Here we go again! After giving it some thought (and taking time to better understand Python packaging) I've adopted most of the changes that you proposed. Moving on to reviewing your own PR!

**Review by dhruvbaldawa — APPROVED** [review]
